contributors/design-proposals/node/node-fence.md

```diff
@@ -0,0 +1,528 @@
+
+## User experience
+The loss of a worker node should be transparent to user's of
+StatefulSets.  Recovery time for affected Pods should be bounded and
+short, allowing scale up/down events to proceed as normal afterwards.
+
+In the absence of this feature, an end-user has no ability to safely
+or reliably allow StatefulSets to be recovered and as such end-users
+will not be provided with a mechanism to enable/disable this
+functionality on a set-by-set basis.
+
+### Use Cases
+1. In a bare metal Kubernetes deployment, StatefulSets should not
+   require admin intervention in order to restore capacity when a
+   member Pod was active on a lost worker
+
+1. In a bare metal Kubernetes deployment, StatefulSets should not
+   require admin intervention in order to scale up or down when a
+   member Pod was active on a lost worker
+
+1. In a Cloud Kubernetes deployment, StatefulSets without an
+   autoscaler should not require admin intervention in order to scale
+   up or down when a member Pod was active on a lost worker
+
+In other words, the failure of a worker node should not represent a
+single point of failure for StatefulSets.
+
+## Admin experience
+### Configurations required
+- How to trigger fence devices/apis - general template parameters (e.g. cluster UPS address and credentials) and overrides values per node for specific fields (e.g. ports related to node)
+- How to run a fence flow for each node (each node can be attached to several fence devices\apis and the fence flow can defer)
+
+The following is done by `ConfigMap` objects:
+
+#### Node Fence Config
+For each node that supports fencing we will configure NodeFenceConfig - This object is needed to centralize the information about “how” the node can be “fenced” from the cluster - we will have 3 steps for fence “isolation”, “power-management”, “recovery”. In each we will have a list of methods to perform.
+
+Note: to keep those fence configuration we had two options - Adding attributes in metadata section to the existing node object or create ConfigMap that describe this information for each node with fencing options.
+```yaml
+kind: ConfigMap
+apiVersion: v1
+metadata:
+name: node_fence_config
+data: 
+  node_fence.properties: | 
+node_name=xxx
+isolation=[method-storage-iso1]]
+power_managment=[method-pm1, method-kdump]
+recovery=[method1,method2,method3]
+execution_policy=[seq\cun]
+```
+Notes:
+- Each configmap we relate a node name - this should be unique, and in the properties we set order for stages’ names.
+- execution_policy means: In list of method we might want to run all methods until all finish, or until only one finishes successfully (useful for sequential pm actions). 
+- Isolation\power_managment\recovery are parameters with list value which represent the method to run in each step of the fencing.
+#### Fence Method Config ###
```

Comment:
This seems like an implementation detail and doesn't relate to the proposal.

Reply:
The method config structure part? But this is the core for working with all those external devices and apis. unless you think that those values need to be stored differently